doc/man/opam.1
.\" Pipe this output to groff -man -Tutf8 | less
.\"
.TH "OPAM" 1 "" "Opam 1.1.2" "Opam Manual"
.\" Disable hyphenantion and ragged-right
.nh
.ad l
.SH NAME
.P
opam \- source\-based OCaml package management
.SH SYNOPSIS
.P
\fBopam\fR \fICOMMAND\fR ...
.SH DESCRIPTION
.P
OPAM is a package manager for OCaml. It uses the powerful mancoosi tools to handle dependencies, including support for version constraints, optional dependencies, and conflict management.
.P
It has support for different remote repositories such as HTTP, rsync, git, darcs and mercurial. It handles multiple OCaml versions concurrently, and is flexible enough to allow you to use your own repositories and packages in addition to the central ones it provides.
.P
Use either \fBopam <command> \-\-help\fR or \fBopam help <command>\fR for more information on a specific command.
.SH COMMANDS
.TP 4
\fBconfig\fR
Display configuration options for packages.
.TP 4
\fBhelp\fR
Display help about OPAM and OPAM commands.
.TP 4
\fBinit\fR
Initialize OPAM state.
.TP 4
\fBinstall\fR
Install a list of packages.
.TP 4
\fBlist\fR
Display the list of available packages.
.TP 4
\fBpin\fR
Pin a given package to a specific version.
.TP 4
\fBreinstall\fR
Reinstall a list of packages.
.TP 4
\fBremove\fR
Remove a list of packages.
.TP 4
\fBrepository\fR
Manage OPAM repositories.
.TP 4
\fBsearch\fR
Search into the package list.
.TP 4
\fBshow\fR
Display information about specific packages.
.TP 4
\fBswitch\fR
Manage multiple installation of compilers.
.TP 4
\fBupdate\fR
Update the list of available packages.
.TP 4
\fBupgrade\fR
Upgrade the installed package to latest version.
.SH COMMANDS ALIASES
.TP 4
\fBinfo\fR
An alias for \fBshow\fR.
.TP 4
\fBremote\fR
An alias for \fBrepository\fR.
.TP 4
\fBuninstall\fR
An alias for \fBremove\fR.
.SH COMMON OPTIONS
.P
These options are common to all commands.
.TP 4
\fB\-\-color\fR=\fIWHEN\fR (absent=auto)
Colorize the output. \fIWHEN\fR must be `always', `never' or `auto'.
.TP 4
\fB\-\-compat\-mode\-1.0\fR
Compatibility mode with OPAM 1.0
.TP 4
\fB\-\-debug\fR
Print debug message on stdout. This is equivalent to setting \fB$OPAMDEBUG\fR to a non\-empty value.
.TP 4
\fB\-\-git\-version\fR
Print the git version if it exists and exit.
.TP 4
\fB\-\-help\fR[=\fIFMT\fR] (default=pager)
Show this help in format \fIFMT\fR (pager, plain or groff).
.TP 4
\fB\-\-no\-aspcud\fR
Do not use the external aspcud solver, even if available.
.TP 4
\fB\-\-no\-base\-packages\fR
Do not install base packages (useful for testing purposes). This is equivalent to setting \fB$OPAMNOBASEPACKAGES\fR to a non\-empty string.
.TP 4
\fB\-q\fR, \fB\-\-quiet\fR
Be quiet when installing a new compiler.
.TP 4
\fB\-r\fR \fIROOT\fR, \fB\-\-root\fR=\fIROOT\fR (absent=~/.opam)
Use \fIROOT\fR as the current root path. This is equivalent to setting \fB$OPAMROOT\fR to \fIROOT\fR.
.TP 4
\fB\-\-strict\fR
Fail whenever an error is found in a package definition or a configuration file. The default is to continue silently if possible.
.TP 4
\fB\-\-switch\fR=\fISWITCH\fR
Use \fISWITCH\fR as the current compiler switch. This is equivalent to setting \fB$OPAMSWITCH\fR to \fISWITCH\fR.
.TP 4
\fB\-v\fR, \fB\-\-verbose\fR
Be more verbose. This is equivalent to setting \fB$OPAMVERBOSE\fR to a non\-empty value.
.TP 4
\fB\-\-version\fR
Show version information.
.TP 4
\fB\-y\fR, \fB\-\-yes\fR
Disable interactive mode and answer yes to all questions that would otherwise be asked to the user. This is equivalent to setting \fB$OPAMYES\fR to a non\-empty string.
.SH ENVIRONMENT VARIABLES
.P
opam makes use of environment variables listed here.
.P
\fIOPAMCOLOR\fR, when set to \fIalways\fR or \fInever\fR, sets a default value for the \-\-color option
.P
\fIOPAMCURL\fR can be used to define an alternative for the 'curl' command\-line utility to download files
.SH FURTHER DOCUMENTATION
.P
See https://opam.ocaml.org.
.SH AUTHORS
.P
Thomas Gazagnaire <thomas@gazagnaire.org>
.sp -1
.P
Anil Madhavapeddy <anil@recoil.org>
.sp -1
.P
Fabrice Le Fessant <Fabrice.Le_fessant@inria.fr>
.sp -1
.P
Frederic Tuong <tuong@users.gforge.inria.fr>
.sp -1
.P
Louis Gesbert <louis.gesbert@ocamlpro.com>
.sp -1
.P
Vincent Bernardoff <vb@luminar.eu.org>
.sp -1
.P
Guillem Rieu <guillem.rieu@ocamlpro.com>
.SH BUGS
.P
Check bug reports at https://github.com/OCamlPro/opam/issues.

CHANGES
1.1.2
* Backport the new OCamlMakefile-based build system from 1.2

1.1.1
* Fix `opam-admin make <packages> -r` (#990)
* Explicitly prettyprint list of lists, to fix `opam-admin depexts` (#997)
* Tell the user which fields is invalid in a configuration file (#1016)
* Add `OpamSolver.empty_universe` for flexible universe instantiation (#1033)
* Add `OpamFormula.eval_relop` and `OpamFormula.check_relop` (#1042)
* Change `OpamCompiler.compare` to match `Pervasives.compare` (#1042)
* Add `OpamCompiler.eval_relop` (#1042)
* Add `OpamPackage.Name.compare` (#1046)
* Add types `version_constraint` and `version_formula` to `OpamFormula` (#1046)
* Clearer command aliases. Made `info` an alias for `show` and added the alias `uninstall` (#944)
* Fixed `opam init --root=<relative path>` (#1047)
* Display OS constraints in `opam info` (#1052)
* Add a new 'opam-installer' script to make `.install` files usable outside of opam (#1026)
* Add a `--resolve` option to `opam-admin make` that builds just the archives you need
  for a specific installation (#1031)
* Fixed handling of spaces in filenames in internal files (#1014)
* Replace calls to `which` by a more portable call (#1061)
* Fixed generation of the init scripts in some cases (#1011)
* Better reports on package patch errors (#987, #988)
* More accurate warnings for unknown package dependencies (#1079)
* Added `opam config report` to help with bug reports (#1034)
* Do not reinstall dev packages with `opam upgrade <pkg>` (#1001)
* Be more careful with `opam init` to a non-empty root directory (#974)
* Cleanup build-dir after successful compiler installation to save on space (#1006)
* Improved OSX compatibility in the external solver tools (#1074)
* Fixed messages printed on update that were plain wrong (#1030)
* Improved detection of meaningful changes from upstream packages to trigger recompilation

1.1.0 [Oct 2013]
* Fix update of dev packages (#962)
* Add support for zip source archives (#958)
* Add `OPAMCURL` environment variable to control invocation of curl (#960)
* Ensure repository redirects only happen for http remotes (#955)
* Turn malformed package files into warnings instead of hard errors (#957)
* Improve robustness of pinned package update (#949)
* Finish conversion of default repository to <https://opam.ocaml.org> (#948)
* Fix regression in handling archives with no extension (treat them as tar again) (#972)
* Fixed stale archives causing packages to be marked as NEW when they weren't (#945)

1.1.0RC1 [Oct 2013]
* Add `make cold` target to build OPAM without a system OCaml installed (#910)
* More informative error messages from `curl` (#905)
* Document use of `OPAMCOLOR` for optional ANSI coloring
* Add `opam-admin depexts` utility to rewrite OPAM files with external dependencies
* Added `repo` files for repository meta-information
* Added support for repo redirections
* Added scripts for automated testing in Travis
* Fixed bug in opam-admin that could keep not up-to-date archives
* Added an `opam-admin depexts` script to ease handling of external dependencies
* Added the `--deps-only` option to `opam install`
* Fixed upgrade with corner-cases of orphan packages
* Added a `note` display form
* Better handling of external solver failures, and added a `--no-aspcud` option
* Fixed unpinning of some installed packages
* Fixed upgrade of metadata from 1.0 when there are orphan custom compilers

1.1.0-beta [Sept 2013]
* Automatic backup before any operation which might alter the list of installed packages
* Support for arbitrary sub-directories for metadata repositories
* Lots of colors
* New option `opam update -u` equivalent to `opam update && opam upgrade --yes`
* New `opam-admin` tool, bundling the features of `opam-mk-repo` and `opam-repo-check` + new 'opam-admin stats' tool
* New `available`: field in opam files, superseding `ocaml-version` and `os` fields
* Package names specified on the command-line are now understood case-insensitively (#705)
* Fixed parsing of malformed opam files (#696)
* Fixed recompilation of a package when uninstalling its optional dependencies (#692)
* Added conditional post-messages support, to help users when a package fails to install for a known reason (#662)
* Rewrite the code which updates pin et dev packages to be quicker and more reliable
* Add {opam,url,desc,files/} overlay for all packages
* `opam config env` now detects the current shell and outputs a sensible default if no override is provided.
* Improve `opam pin` stability and start display information about dev revisions
* Add a new `man` field in `.install` files
* Support hierarchical installation in `.install` files
* Add a new `stublibs` field in `.install` files
* OPAM works even when the current directory has been deleted
* speed-up invocation of `opam config var VARIABLE` when variable is simple (eg. `prefix`, `lib`, ...)
* `opam list` now display only the installed packages. Use `opam list -a` to get the previous behavior.
* Inverse the depext tag selection (useful for `ocamlot`)
* Add a `--sexp` option to `opam config env` to load the configuration under emacs
* Purge `~/.opam/log` on each invocation of OPAM
* System compiler with versions such as `version+patches` are now handled as if this was simply `version`
* New `OpamVCS` functor to generate OPAM backends
* More efficient `opam update`
* Switch license to LGPL with linking exception
* `opam search` now also searches through the tags
* minor API chanages for `API.list` and `API.SWITCH.list`
* Improve the syntax of filters
* Add a `messages` field
* Add a `--jobs` command line option and add `%{jobs}%` to be used in OPAM files
* Various improvments in the solver heuristics
* By default, turn-on checking of certificates for downloaded dependency archives: use `./configure --disable-certificate-check` to go back to the previous behavior
* Check the md5sum of downloaded archives when compiling OPAM

src/opam.ml
(***********************************************************************)
(*                                                                     *)
(*    Copyright 2012 OCamlPro                                          *)
(*    Copyright 2012 INRIA                                             *)
(*                                                                     *)
(*  All rights reserved.  This file is distributed under the terms of  *)
(*  the GNU Public License version 3.0.                                *)
(*                                                                     *)
(*  OPAM is distributed in the hope that it will be useful,            *)
(*  but WITHOUT ANY WARRANTY; without even the implied warranty of     *)
(*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      *)
(*  GNU General Public License for more details.                       *)
(*                                                                     *)
(***********************************************************************)

open Utils
open Types
open Path
open Solver
open Client
open SubCommand

let ano_args = ref []
let anon s =
  ano_args := s :: !ano_args

exception Bad of string * string

let bad_argument cmd fmt =
  Printf.ksprintf (fun msg -> raise (Bad (cmd, msg))) fmt

let noanon cmd s =
  raise (Bad (cmd, s ^ " is not expected"))

let () = Globals.root_path := Globals.default_opam_path

(* Useful for switch, which can overwrite the default verbose flag *)
let quiet = ref false

let global_args = [
  "--debug"     , Arg.Set Globals.debug   , " Print internal debug messages (very verbose)";
  "--verbose"   , Arg.Set Globals.verbose , " Display the output of subprocesses";
  "--quiet"     , Arg.Clear quiet         , " Do not display the output of subprocesses";
  "--version"   , Arg.Unit Globals.version_msg, " Display version information";
  "--yes"       , Arg.Set Globals.yes     , " Answer yes to all questions";
  "--makecmd"   , Arg.Set_string Globals.makecmd,
    Printf.sprintf " Set the 'make' program used when compiling packages (default is %s)" !Globals.makecmd;
  "--root"      , Arg.Set_string Globals.root_path,
    (Printf.sprintf " Change root path (default is %s)" Globals.default_opam_path);
  "--no-checksums", Arg.Clear Globals.verify_checksums, " Do not verify checksums on download";
]

let parse_args fn () =
  fn (List.rev !ano_args)

let guess_repository_kind kind address =
  match kind with
  | None  ->
      if Sys.file_exists address then
        "rsync"
      else if Utils.starts_with ~prefix:"git" address
           || Utils.ends_with ~suffix:"git" address then
        "git"
      else
        Globals.default_repository_kind
  | Some k -> k

(* opam init [-kind $kind] $repo $adress *)
let init =
  let kind = ref None in
  let comp = ref None in
  let cores = ref Globals.default_cores in
{
  name     = "init";
  usage    = "";
  synopsis = "Initial setup";
  help     = "Create the initial config files";
  specs    = [
    ("-comp" , Arg.String (fun s -> comp := Some (OCaml_V.of_string s)), " Which compiler version to use");
    ("-cores", Arg.Set_int cores   , " Set the number of cores");
    ("-kind" , Arg.String (fun s -> kind := Some s) , " Set the repository kind");
    ("-no-base-packages", Arg.Clear Globals.base_packages, " Do not install the base packages");
  ];
  anon;
  main     =
    parse_args (function
    | [] ->
        Client.init Repository.default !comp !cores
    | [address] ->
        let name = Globals.default_repository_name in
        let kind = guess_repository_kind !kind address in
        let repo = Repository.create ~name ~address ~kind in
        Client.init repo !comp !cores
    | [name; address] ->
        let kind = guess_repository_kind !kind address in
        let repo = Repository.create ~name ~address ~kind in
        Client.init repo !comp !cores
    | _ -> bad_argument "init" "Need a repository name and address")
}

(* opam list [PACKAGE_REGEXP]* *)
let list =
  let print_short = ref false in
  let installed_only = ref false in
{
  name     = "list";
  usage    = "<package-regexp>*";
  synopsis = "Display the list of available packages";
  help     = "";
  specs    = [
    ("-short"    , Arg.Set print_short   , " Minimize the output by displaying only package name");
    ("-installed", Arg.Set installed_only, " Display only the list of installed packages");
  ];
  anon;
  main     =
    parse_args (function args ->
      let print_short = !print_short in
      let installed_only = !installed_only in
      Client.list ~print_short ~installed_only args
    )
}

(* opam search [PACKAGE_REGEXP]* *)
let search =
  let print_short = ref false in
  let installed_only = ref false in
  let case_sensitive = ref false in
{
  name     = "search";
  usage    = "<package-regexp>*";
  synopsis = "Search into the package list";
  help     = "";
  specs    = [
    ("-short"    , Arg.Set print_short   , " Minimize the output by displaying only package name");
    ("-installed", Arg.Set installed_only, " Display only the list of installed packages");
    ("-case-sensitive", Arg.Set case_sensitive, " Force the search in case sensitive (insensitive by default)");
  ];

  anon;
  main     =
    parse_args (function args ->
      let print_short = !print_short in
      let installed_only = !installed_only in
      let case_sensitive = !case_sensitive in
      Client.list ~print_short ~installed_only ~name_only:false ~case_sensitive args
    )
}

(* opam info [PACKAGE] *)
let info = {
  name     = "info";
  usage    = "[package]+";
  synopsis = "Display information about specific packages";
  help     = "";
  specs    = [];
  anon;
  main     =
    parse_args (function
    | [] -> bad_argument "info" "Missing package argument"
    | l  -> List.iter (fun name -> Client.info (N.of_string name)) l)
}

(* opam config [-r [-I|-bytelink|-asmlink] PACKAGE+ *)
let config =
let has_cmd = ref false in
let is_rec = ref false in
let is_link = ref false in
let is_byte = ref false in
let bytecomp () =  has_cmd := true; is_byte := true ; is_link := false in
let bytelink () =  has_cmd := true; is_byte := true ; is_link := true in
let asmcomp  () =  has_cmd := true; is_byte := false; is_link := false in
let asmlink  () =  has_cmd := true; is_byte := false; is_link := true in
let command = ref None in
let set cmd () =
  has_cmd := true;
  command := Some cmd in
let specs = [
  ("-r"        , Arg.Set is_rec       , " Recursive search");
  ("-I"        , Arg.Unit (set `I)    , " Display include options");
  ("-bytecomp" , Arg.Unit bytecomp    , " Display bytecode compile options");
  ("-asmcomp"  , Arg.Unit asmcomp     , " Display native compile options");
  ("-bytelink" , Arg.Unit bytelink    , " Display bytecode link options");
  ("-asmlink"  , Arg.Unit asmlink     , " Display native link options");
  ("-list-vars", Arg.Unit (set `List) , " Display the contents of all available variables");
  ("-var"      , Arg.Unit (set `Var)  , " Display the content of a variable");
  ("-subst"    , Arg.Unit (set `Subst), " Substitute variables in files");
  ("-env"      , Arg.Unit (set `Env)  , " Display the compiler environment variables");
] in
let mk options =
  if not !has_cmd then
    bad_argument "config"
      "Wrong options (has_cmd=%b is_rec=%b,is_link=%b,is_byte=%b)"
      !has_cmd !is_rec !is_link !is_byte
  else
    Compil {
      is_rec  = !is_rec;
      is_link = !is_link;
      is_byte = !is_byte;
      options = List.map Full_section.of_string options;
    } in
{
  name     = "config";
  usage    = "[...]+";
  synopsis = "Display configuration options for packages";
  help     = "";
  specs;
  anon;
  main = function () ->
    let names = List.rev !ano_args in
    let config = match !command with
      | Some `Env   -> Env
      | Some `I     -> Includes (!is_rec, List.map N.of_string names)
      | Some `List  -> List_vars
      | Some `Var when List.length names = 1
                    -> Variable (Full_variable.of_string (List.hd names))
      | Some `Var   ->
          bad_argument "config" "-var takes exactly one parameter"
      | Some `Subst -> Subst (List.map Basename.of_string names)
      | None        -> mk names in
    Client.config config
}

(* opam install <PACKAGE>+ *)
let install = {
  name     = "install";
  usage    = "[package]+";
  synopsis = "Install a list of packages";
  help     = "";
  specs    = [];
  anon;
  main     = parse_args (fun names ->
    if names <> [] then
      let names = List.map N.of_string names in
      Client.install (N.Set.of_list names)
    else Globals.error_and_exit "You need to specify at least one package to install."
  )
}

(* opam reinstall <PACKAGE>+ *)
let reinstall = {
  name     = "reinstall";
  usage    = "[package]+";
  synopsis = "Reinstall a list of packages";
  help     = "";
  specs    = [];
  anon;
  main     = parse_args (fun names ->
    let names = List.map N.of_string names in
    Client.reinstall (N.Set.of_list names)
  )
}

(* opam update *)
let update = {
  name     = "update";
  usage    = "[repo]*";
  synopsis = "Update the list of available packages";
  help     = "";
  specs    = [];
  anon;
  main     = parse_args Client.update
}

(* opam upgrade *)
let upgrade = {
  name     = "upgrade";
  usage    = "[package]*";
  synopsis = "Upgrade the installed package to latest version";
  help     = "";
  specs    = [];
  anon;
  main     = parse_args (fun names ->
    let names = List.map N.of_string names in
    Client.upgrade (N.Set.of_list names);
  )
}

(* opam upload PACKAGE *)
let upload =
  let opam = ref "" in
  let descr = ref "" in
  let archive = ref "" in
  let repo = ref "" in
{
  name     = "upload";
  usage    = "";
  synopsis = "Upload a package to an OPAM repository";
  help     = "";
  specs    = [
    ("-opam"   , Arg.Set_string opam   , " specify the OPAM file to upload");
    ("-descr"  , Arg.Set_string descr  , " specify the description file to upload");
    ("-archive", Arg.Set_string archive, " specify the archive file to upload");
    ("-repo"   , Arg.Set_string repo   , " (optional) specify the repository to upload")
  ];
  anon = noanon "upload";
  main     = (function () ->
    if !opam = "" then
      bad_argument "upload" "missing OPAM file";
    if !descr = "" then
      bad_argument "upload" "missing description file";
    if !archive = "" then
      bad_argument "upload" "missing archive file";
    let opam = Filename.of_string !opam in
    let descr = Filename.of_string !descr in
    let archive = Filename.of_string !archive in
    let repo = if !repo = "" then None else Some !repo in
    Client.upload { opam; descr; archive } repo)
}

(* opam remove PACKAGE *)
let remove = {
  name     = "remove";
  usage    = "[package]+";
  synopsis = "Remove a list of packages";
  help     = "";
  specs    = [];
  anon;
  main     = parse_args (fun names ->
    if names <> [] then
      let names = List.map N.of_string names in
      Client.remove (N.Set.of_list names)
    else Globals.error_and_exit "You need to specify at least one package to remove."
  )
}

(* opam remote [-list|-add <url>|-rm <url>] *)
let remote =
  let kind = ref None in
  let command : [`add|`list|`rm] option ref = ref None in
  let set c () = command := Some c in
{
  name     = "remote";
  usage    = "[-list|add <name> <address>|rm <name>]";
  synopsis = "Manage remote servers";
  help     = "";
  specs    = [
    ("-list" , Arg.Unit (set `list), " List the repositories");
    ("-add"  , Arg.Unit (set `add) , " Add a new repository");
    ("-rm"   , Arg.Unit (set `rm)  , " Remove a remote repository");
    ("-kind" , Arg.String (fun s -> kind := Some s) , " (optional) Specify the repository kind");
  ];
  anon;
  main     = parse_args (fun args ->
    match !command, args with
    | Some `list, []                -> Client.remote List
    | Some `rm,   [ name ]          -> Client.remote (Rm name)
    | Some `add , [ name; address ] ->
        let kind = guess_repository_kind !kind address in
        Client.remote (Add (Repository.create ~name ~kind ~address))
    | None, _  -> bad_argument "remote" "Command missing [-list|-add|-rm]"
    | _        -> bad_argument "remote" "Wrong arguments")
}

(* opam switch [-clone] OVERSION *)
let switch =
  let alias_of = ref "" in
  let command = ref `switch in
  let set c () =
    if !command <> `switch then
      bad_argument "switch" "two many sub-commands";
    command := c in
  let no_alias_of () =
    if !alias_of <> "" then
      bad_argument "switch" "invalid -alias-of option" in
  let mk_comp alias = match !alias_of with
    | ""   -> OCaml_V.of_string alias
    | comp -> OCaml_V.of_string comp in
{
  name     = "switch";
  usage    = "[compiler-name]";
  synopsis = "Manage multiple installation of compilers";
  help     = "";
  specs    = [
    ("-alias-of"        , Arg.Set_string alias_of        , " Compiler name");
    ("-no-base-packages", Arg.Clear Globals.base_packages, " Do not install the base packages");
    ("-install"         , Arg.Unit (set `install)        , " Install the given compiler");
    ("-remove"          , Arg.Unit (set `remove)         , " Remove the given compiler");
    ("-export"          , Arg.String (fun s -> set (`export s) ()), " Export the libraries installed with the given alias");
    ("-import"          , Arg.String (fun s -> set (`import s) ()), " Import the libraries installed with the given alias");
    ("-reinstall"       , Arg.Unit (set `reinstall)      , " Reinstall the given compiler");
    ("-list"            , Arg.Unit (set `list)           , " List the available compilers");
    ("-current"         , Arg.Unit (set `current)        , " Display the current compiler");
  ];
  anon;
  main     = parse_args (function args ->
    match !command, args with
    | `install, [alias] ->
        Client.compiler_install !quiet (Alias.of_string alias) (mk_comp alias)
    | `export f, [] ->
        no_alias_of ();
        Client.compiler_export (Filename.of_string f)
    | `import f, [] ->
        no_alias_of ();
        Client.compiler_import (Filename.of_string f)
    | `remove, aliases ->
        no_alias_of ();
        List.iter (fun alias -> Client.compiler_remove (Alias.of_string alias)) aliases
    | `reinstall, [alias] ->
        no_alias_of ();
        Client.compiler_reinstall (Alias.of_string alias)
    | `list, [] ->
        no_alias_of ();
        Client.compiler_list ()
    | `current, [] ->
        no_alias_of ();
        Client.compiler_current ()
    | `switch, [alias] ->
        begin match !alias_of with
          | "" -> Client.compiler_switch !quiet (Alias.of_string alias)
          | _  -> Client.compiler_install !quiet (Alias.of_string alias) (mk_comp alias)
        end
    | _ -> bad_argument "switch" "too many arguments"
  )
}

(* opam pin [-list|<package> <version>|<package> <path>] *)
let pin =
  let list = ref false in
  let kind = ref None in
  let set_kind s = kind := Some s in
{
  name     = "pin";
  usage    = "<package> [<version>|<url>|none]";
  synopsis = "Pin a given package to a specific version";
  help     = "";
  specs    = [
    ("-list", Arg.Set list       , " List the current status of pinned packages");
    ("-kind", Arg.String set_kind, " Force the pin action (options are: 'git', 'rsync', 'version'");
  ];
  anon;
  main     = parse_args (function
    | [] when !list    -> Client.pin_list ()
    | [name; arg]      -> Client.pin { pin_package = N.of_string name; pin_arg = pin_option_of_string ?kind:!kind arg }
    | _                -> bad_argument "pin" "Wrong arguments")
}

let commands = [
  init;
  list;
  search;
  info;
  config;
  install;
  reinstall;
  update;
  upgrade;
  upload;
  remove;
  remote;
  switch;
  pin;
]

let () =
  Sys.catch_break true;
  Printexc.register_printer (function
    | Unix.Unix_error (e,fn, msg) ->
      let msg = if msg = "" then "" else " on " ^ msg in
      let error = Printf.sprintf "%s: %S failed%s: %s" Sys.argv.(0) fn msg (Unix.error_message e) in
      Some error
    | _ -> None);
  List.iter SubCommand.register commands;
  try ArgExt.parse ~man_fun:
        (fun cmd -> ignore (Sys.command ("man opam-" ^ cmd))) global_args
  with e ->
    Globals.error "  '%s' failed\n" (String.concat " " (Array.to_list Sys.argv));
    match e with
    | Bad (cmd, msg) ->
        ArgExt.pp_print_help (ArgExt.SubCommand cmd) Format.err_formatter global_args ();
        Globals.error "%s" msg;
        exit 1;
    | Failure ("no subcommand defined" as s) ->
        ArgExt.pp_print_help ArgExt.NoSubCommand Format.err_formatter global_args ();
        Globals.error "%s" s;
        exit 2
    | Globals.Exit i -> exit i
    | e ->
      let bt = Printexc.get_backtrace () in
      let bt = if bt = "" then "" else Printf.sprintf "    at\n %s\n" bt in
      Printf.fprintf stderr "Fatal error: exception %s\n%s%!"
        (Printexc.to_string e) bt;
      exit 2
